test/timeline_test.dart

```diff
@@ -0,0 +1,148 @@
+
+  TestClient(api.ClientChannel channel) : super(channel);
+  ResponseStream<int> stream(int request, {CallOptions options}) {
+    final call =
+        $createCall(_$stream, Stream.fromIterable([request]), options: options);
+    return ResponseStream(call);
+  }
+}
+
+class TestService extends Service {
+  String get $name => 'test.TestService';
+
+  TestService() {
+    $addMethod(ServiceMethod<int, int>('stream', stream, false, true,
+        (List<int> value) => value[0], (int value) => [value]));
+  }
+
+  Stream<int> stream(ServiceCall call, Future request) async* {
+    yield 1;
+    yield 2;
+    yield 3;
+  }
+}
+
+class FixedConnectionClientChannel extends ClientChannelBase {
+  final Http2ClientConnection clientConnection;
+  List<ConnectionState> states = <ConnectionState>[];
+  FixedConnectionClientChannel(this.clientConnection) {
+    clientConnection.onStateChanged = (c) => states.add(c.state);
+  }
+  @override
+  ClientConnection createConnection() => clientConnection;
+}
+
+testee() async {
+  final Server server = Server([TestService()]);
+  await server.serve(address: 'localhost', port: 0);
+
+  enableTimelineLogging();
+  final channel = FixedConnectionClientChannel(Http2ClientConnection(
+    'localhost',
+    server.port,
+    ChannelOptions(credentials: ChannelCredentials.insecure()),
+  ));
+  final testClient = TestClient(channel);
+  await testClient.stream(1).toList();
+  await server.shutdown();
+}
+
+void checkStartEvent(List<Map> events) {
+  final e = events.firstWhere((e) => e['ph'] == 'b');
+  expect(e['args']['method'], isNotNull);
+  expect(e['args']['method'], equals(path));
+}
+
+void checkSendEvent(List<Map> events) {
+  final e = events.firstWhere((e) => e['name'] == 'Request sent');
+  expect(e['args']['metadata'], isNotNull);
+}
+
+void checkWriteEvent(List<Map> events) {
+  final e = events.firstWhere((e) => e['name'] == 'Data sent');
+  expect(e['args']['data'], isNotNull);
+  expect(e['args']['data'], equals('1'));
+}
+
+void checkReceiveEvent(List<Map> events) {
+  events = events.where((e) => e['name'] == 'Data received').toList();
+  expect(events.length, equals(3));
+  int sum = 0;
+  for (final e in events) {
+    // 3 elements are 1, 2 and 3.
+    sum |= 1 << int.parse(e['args']['data']);
+  }
+  expect(sum, equals(14));
+}
+
+void checkReceiveMetaDataEvent(List<Map> events) {
+  events = events.where((e) => e['name'] == 'Metadata received').toList();
+  expect(events.length, equals(2));
+  for (final e in events) {
+    if (e['args']['headers'] != null) {
+      final header = e['args']['headers'];
+      expect(header, contains('status: 200'));
+      expect(header, contains('content-type: application/grpc'));
+    } else {
+      expect(e['args']['trailers'], contains('grpc-status: 0'));
+    }
+  }
+}
+
+void checkFinishEvent(List<Map> events) {
+  final e = events.firstWhere((e) => e['ph'] == 'e');
+  expect(e, isNotNull);
+}
+
+var tests = <IsolateTest>[
```

Comment:
Yeah, I don't think we need to pull in the vm_service test harness just to setup a `VmService` instance. Just pull out the relevant code you need since we probably won't be writing many tests like this in this repo.

Reply:
I've already removed most of unused codes. But it still has more than 300 lines of code. So I put them in a separate file.

Reply:
Sure, but I don't think we want to have all the extra overhead of the options we're not using. I'd remove everything that you don't explicitly need.

Reply:
Done. I only kept functions that are used in the test in the 'test_util.dart'.

Reply:
A simpler way to test this would be to provide a mock implementation of `TimelineTask` (e.g. in the `profile.dart` you can have a configurable factory for timeline tasks) then you would not need to setup the whole vmservice stuff in here at all and instead just test against the mock. I leave it up to you to decide if you want to simplify this.